lib/nerves_hub_web/channels/device_socket_token_auth.ex

```diff
@@ -0,0 +1,111 @@
+defmodule NervesHubWeb.DeviceSocketTokenAuth do
+  use Phoenix.Socket
+
+  alias NervesHub.Devices
+
+  alias Plug.Crypto
+
+  channel("console", NervesHubWeb.ConsoleChannel)
+  channel("device", NervesHubWeb.DeviceChannel)
+
+  @salt_headers [
+    "x-nh-key-digest",
+    "x-nh-key-iterations",
+    "x-nh-key-length",
+    "x-nh-access-id",
+    "x-nh-time"
+  ]
+
+  # Default 15 min max age for the signature
+  @default_max_age 900
+
+  def connect(_params, socket, %{x_headers: headers}) do
+    parsed_data = parse_headers(headers)
+    verify_opts = verify_options(parsed_data)
+    salt = expected_salt(headers)
+
+    with {:ok, true} <- Application.fetch_env(:nerves_hub, __MODULE__)[:enabled],
+         {:ok, access_id} <- Keyword.fetch(parsed_data, :access_id),
+         {:ok, token_auth} <- get_product_token_auth(access_id),
+         {:ok, signature} <- Keyword.fetch(parsed_data, :signature),
+         {:ok, identifier} <- Crypto.verify(token_auth.secret, salt, signature, verify_opts),
+         {:ok, device} <-
+           Devices.get_or_create_device(token_auth: token_auth, identifier: identifier) do
+      socket =
+        socket
+        |> assign(:device, device)
+        |> assign(:reference_id, generate_reference_id())
+
+      {:ok, socket}
+    else
+      _ -> {:error, :invalud_auth}
+    end
+  end
+
+  def connect(_params, _socket, _connect_info), do: :error
+
+  def id(%{assigns: %{device: device}}), do: "device_socket:#{device.id}"
+  def id(_socket), do: nil
+
+  defp parse_headers(headers) do
+    for {k, v} <- headers do
+      case String.downcase(k) do
+        "x-nh-time" ->
+          {:signed_at, String.to_integer(v)}
+
+        "x-nh-key-length" ->
+          {:key_length, String.to_integer(v)}
+
+        "x-nh-key-iterations" ->
+          {:key_iterations, String.to_integer(v)}
+
+        "x-nh-key-digest" ->
+          "NH1-HMAC-" <> digest_str = v
+          {:key_digest, String.to_existing_atom(String.downcase(digest_str))}
+
+        "x-nh-access-id" ->
+          {:access_id, v}
+
+        "x-nh-signature" ->
+          {:signature, v}
+
+        _ ->
+          # TODO: What should we do with this unknown?
+          nil
+      end
+    end
+  end
+
+  defp verify_options(parsed_data) do
```

Comment:
```suggestion
  defp verification_opts(parsed_data) do
```
Just a nit, but `verify_opts` makes me think we're verifying the options are okay before we use them